### PR TITLE
Easily understandable error if inference steps not set before using scheduler (#263)

### PR DESCRIPTION
**Pull request by samedii**
Ran into this when building around diffusers. Did this in the browser but tested locally first.

P.S. Great refactor of the original LDM code!

### CODE REVIEW

**Comment by pcuenca** [comment]
Hi @samedii, thanks a lot for writing!

The inference steps don't need to match the number of training steps. I believe the intention is to configure your scheduler for inference using [`set_timesteps()`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_ddpm.py#L90).

In the example pipeline for ddpm, the number of inference steps is indeed set to `1000`: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/ddpm/pipeline_ddpm.py#L54

However, in other pipelines such as the one for [Stable Diffusion](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/ddpm/pipeline_ddpm.py#L54) it is set to a smaller number such as `50`.

In general, we provide the official pipelines as a guideline to show how the building blocks can be used together, although you are free or course to use them however you like :)

Please, let us know if that helps :)

**Comment by samedii** [comment]
Hi!
I think there has been a misunderstanding, I know that the number of steps do not need to match. I would have just liked to have a default number of steps. It is now just raising an exception without telling the user why.

If you do not like having a default number of steps you could instead add a check like:
```
if self.num_inference_steps is None:
    raise SomeException("Number of inference steps is 'None', you need to run 'set_timesteps' after creating the scheduler").
```

**Review by anton-l — COMMENTED** [review]
This is a valid addition IMO, the number of inference steps can be set to either a full timescale or a reasonable default like 50. I'm leaning towards 50 though, since 1000 doesn't really showcase the scheduler if the user forgets to set the number of steps initially. cc @patrickvonplaten @patil-suraj

**Comment by patrickvonplaten** [comment]
Very cool addition @samedii ! Agree with @anton-l to set the default to 50 :-)

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 1]
```diff
@@ -84,7 +84,7 @@ def __init__(
         self.final_alpha_cumprod = np.array(1.0) if set_alpha_to_one else self.alphas_cumprod[0]
 
         # setable values
-        self.num_inference_steps = None
+        self.num_inference_steps = num_train_timesteps
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:87]:
```suggestion
        # following the original paper: https://arxiv.org/abs/2010.02502
        # 50 inference steps are often a good trade-off between quality and speed
        # see Table 1 row 1. If you would like to change the number of inference steps
        # please make use the `set_timesteps(num_inference_steps=1000)`
        self.num_inference_steps = 50
```

**Review by patrickvonplaten — APPROVED** [review]

**Comment by samedii** [comment]
Do you want to have a default for every scheduler? Should the default then be the same budget for all (50)?

**Comment by samedii** [comment]
Using `set_timesteps` in `__init__` unexpectedly created a new issue with `_internal_dict` so I removed that suggestion and the changes only concern DDIM again now.

**Review by patil-suraj — COMMENTED** [review]
Thanks for the PR @samedii !

This ideed makes sense, but it introduces some in-consistency here.  Because we are setting the `num_inference_steps` but the `self.timesteps` are still according to `num_train_timesteps`. So the users will have to call `set_timesteps` even if we set a default value.

Now, if we modify the `self.timesteps` according to inference steps in the `__init__` then one can not use this for training. 

Think here we need to decide something first. By default do we want to keep the scheduler in inference mode or training mode ? Depending on that we can set the `self.timestep`
wdyt @anton-l @patrickvonplaten

**Comment by anton-l** [comment]
@patil-suraj one way to solve that would be to set the default `num_inference_steps` inside `step()`, but on the other hand it's not good to to have a magic number 50 outside of `__init__`.  In this situation having 1000 might be better indeed

**Comment by natolambert** [comment]
If we change this, we should definitely change all of them. I also commented on the original issue, but I kind of lean towards forced setting, or in something like a pipeline letting the user load one with the correct number of inference steps.

Generally I think if the user is making an inference loop, needing to set the timesteps mitigates risk of the user having a mismatch between their timesteps and the number of iterations.

**Comment by patil-suraj** [comment]
> @patil-suraj one way to solve that would be to set the default `num_inference_steps` inside `step()`, but on the other hand it's not good to to have a magic number 50 outside of `__init__`. In this situation having 1000 might be better indeed

Yeah, having it inside `step` is not ideal. Maybe let's go with the other solution that @samedii suggested here 
https://github.com/huggingface/diffusers/pull/264#issuecomment-1230027781

raise an error inside `step`

```python
if self.num_inference_steps is None:
    raise SomeException("Number of inference steps is 'None', you need to run 'set_timesteps' after creating the scheduler")
```

**Comment by patrickvonplaten** [comment]
Answered also here: https://github.com/huggingface/diffusers/issues/263

Thinking more about it, I'm also heavily in favor of not giving any default - instead we should go for:

```py
if self.num_inference_steps is None:
    raise ValueError("Number of inference steps is 'None', you need to run 'set_timesteps' after creating the scheduler")
```

just like @patil-suraj suggested here: https://github.com/huggingface/diffusers/pull/264#issuecomment-1230535087 . The number of inference steps is not related to the config of the scheduler and should therefore not be set at init. Sorry should have taken a better look before. 

@samedii - also I'm curious though exactly how you got this error because the pipelines should all automatically call `set_timesteps(...)`

**Comment by samedii** [comment]
> @samedii - also I'm curious though exactly how you got this error because the pipelines should all automatically call `set_timesteps(...)`

Followed the instructions in the main README but adapted it to use DDIM. I don't have a typical use case though, I'm building a lib on top of diffusers.

```
lms = LMSDiscreteScheduler(
    beta_start=0.00085, 
    beta_end=0.012, 
    beta_schedule="scaled_linear"
)

pipe = StableDiffusionPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-3", 
    scheduler=lms,
    use_auth_token=True
).to("cuda")
```

Changed the PR to raising an exception for all schedulers.

File: tests/test_scheduler.py [hunk 2]
```diff
@@ -203,6 +203,7 @@ def test_pytorch_equal_numpy(self):
 
 class DDPMSchedulerTest(SchedulerCommonTest):
     scheduler_classes = (DDPMScheduler,)
+    forward_default_kwargs = (("num_inference_steps", 50),)
```
Comment (patrickvonplaten) [inline tests/test_scheduler.py:206]: DDPM should not be able to change the inference steps IMO -> that's what we have DDIM for :-)

Comment (patrickvonplaten) [inline tests/test_scheduler.py:206]:
```suggestion
```

File: tests/test_scheduler.py [hunk 3]
```diff
@@ -265,6 +266,7 @@ def test_full_loop_no_noise(self):
         model = self.dummy_model()
         sample = self.dummy_sample_deter
 
+        scheduler.set_timesteps(num_trained_timesteps)
```
Comment (patrickvonplaten) [inline tests/test_scheduler.py:269]:
```suggestion
```

**Review by patrickvonplaten — APPROVED** [review]

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 4]
```diff
@@ -136,6 +136,11 @@ def step(
         predict_epsilon=True,
         generator=None,
     ):
+        if self.num_inference_steps is None:
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm.py:139]: Here I don't think we need the error message as DDPM usually runs with number of training steps

Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm.py:139]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 5]
```diff
@@ -136,6 +136,11 @@ def step(
...
+            raise ValueError(
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm.py:140]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 6]
```diff
@@ -136,6 +136,11 @@ def step(
...
+                "Number of inference steps is 'None', you need to run 'set_timesteps' after creating the scheduler"
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm.py:141]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 7]
```diff
@@ -136,6 +136,11 @@ def step(
...
+            )
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm.py:142]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_karras_ve.py [hunk 8]
```diff
@@ -103,6 +103,11 @@ def step(
         sigma_prev: float,
         sample_hat: Union[torch.FloatTensor, np.ndarray],
     ):
+        if self.num_inference_steps is None:
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_karras_ve.py:106]: same here I don't think we should throw an error if we don't have to - I'd be in favor of removing it here

File: src/diffusers/schedulers/scheduling_karras_ve.py [hunk 9]
```diff
@@ -118,6 +123,11 @@ def step_correct(
         sample_prev: Union[torch.FloatTensor, np.ndarray],
         derivative: Union[torch.FloatTensor, np.ndarray],
     ):
+        if self.num_inference_steps is None:
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_karras_ve.py:126]: Also no need for a hard-error here

File: src/diffusers/schedulers/scheduling_lms_discrete.py [hunk 10]
```diff
@@ -101,6 +101,11 @@ def step(
         sample: Union[torch.FloatTensor, np.ndarray],
         order: int = 4,
     ):
+        if self.num_inference_steps is None:
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_lms_discrete.py:104]: same here no need for a hard error

File: src/diffusers/schedulers/scheduling_pndm.py [hunk 11]
```diff
@@ -130,6 +130,11 @@ def step(
         timestep: int,
         sample: Union[torch.FloatTensor, np.ndarray],
     ):
+        if self.num_inference_steps is None:
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_pndm.py:133]: Think it's good enough to only throw the error in the "sub"-step methods

File: src/diffusers/schedulers/scheduling_sde_ve.py [hunk 12]
```diff
@@ -155,6 +160,11 @@ def step_correct(
         if seed is not None:
             self.set_seed(seed)
 
+        if self.timesteps is None:
+            raise ValueError(
+                "Number of inference steps is not set, you need to run 'set_timesteps' after creating the scheduler"
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_sde_ve.py:165]:
```suggestion
                "`self.timesteps` is not set, you need to run 'set_timesteps' after creating the scheduler"
```

File: src/diffusers/schedulers/scheduling_sde_vp.py [hunk 13]
```diff
@@ -35,6 +35,11 @@ def set_timesteps(self, num_inference_steps):
         self.timesteps = torch.linspace(1, self.config.sampling_eps, num_inference_steps)
 
     def step_pred(self, score, x, t):
+        if self.timesteps is None:
+            raise ValueError(
+                "Number of inference steps is not set, you need to run 'set_timesteps' after creating the scheduler"
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_sde_vp.py:40]:
```suggestion
                "`self.timesteps` is not set, you need to run 'set_timesteps' after creating the scheduler"
```

**Comment by samedii** [comment]
Good points! I was a little surprised to see `set_timesteps` in DDPM

**Comment by patrickvonplaten** [comment]
Thanks a lot for iterating with us here! The PR looks good to me - @patil-suraj @anton-l what do you think? :-)

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for @samedii , LGTM!

**Pull request merged**